### ISSUE TITLE
RF06 mutually exclusive with RF04

### ISSUE THREAD

**Issue by ifm-anowaczek** [sqlfluff/sqlfluff#7072]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

`TITLE` is on the list of unreserved keywords for the Snowflake Dialect. Given:
```sql
select foo as title from bar
```
You will fail with `RF04 | Keywords should not be used as identifiers.`. Quoting the keyword like:
```sql
select foo as "TITLE" from bar
```
results in a fail due to `RF06 | Unnecessary quoted identifier "TITLE".`.

Perhaps quoting the keyword is not an intended solution to RF04, but I would not expect to be required to change the name of my identifier because it is the same as an _unreserved_ keyword - particularly in the case of common potential identifiers like `title`, `email`, `password`, etc.

### Expected Behaviour

Quoting `TITLE` should not be seen as "unnecessary".

### Observed Behaviour

Quoting `TITLE` is seen as unnecessary.

### How to reproduce

`sqlfluff lint/fix` on the above code snippets.

### Dialect

Snowflake

### Version

`sqlfluff`: 3.4.1
`Python`: 3.13.2
`sqlfluff-templater-dbt`: 3.4.1
`dbt-core`: 1.10.3
`dbt-snowflake`: 1.9.4

### Configuration

```ini
[tool.sqlfluff.core]
dialect = "snowflake"
templater = "dbt"
max_line_length = 120
exclude_rules = ["ST06", "CV03"]

[tool.sqlfluff.templater.dbt]
profile = "xxxxxxxxxxxx"
target = "xxxxx"
apply_dbt_builtins = true
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by ifm-anowaczek** [comment]
I looked into it a little more and this is happening because in RF06, only reserved keywords are deemed as necessary to be quoted, but in RF04 unreserved keywords fail if not quoted.

[This conditional](https://github.com/sqlfluff/sqlfluff/blob/6e15925fc046e1df4841b477df0effc637e3a3ef/src/sqlfluff/rules/references/RF06.py#L262-L267) excepts reserved keywords from being unquoted:
```py
if regex.fullmatch(
    anti_template,
    identifier_contents,
    regex.IGNORECASE,
):
    return None
```
based on [this pattern](https://github.com/sqlfluff/sqlfluff/blob/3e879b1f64fc69d649faf8bd8d302efb73ebceb8/src/sqlfluff/dialects/dialect_snowflake.py#L625-L636) which includes `reserved_keywords` but not `unreserved_keywords`
```py
snowflake_dialect.replace(
    NakedIdentifierSegment=SegmentGenerator(
        # Generate the anti template from the set of reserved keywords
        lambda dialect: RegexParser(
            # See https://docs.snowflake.com/en/sql-reference/identifiers-syntax.html
            r"[a-zA-Z_][a-zA-Z0-9_$]*",
            IdentifierSegment,
            type="naked_identifier",
            anti_template=r"^(" + r"|".join(dialect.sets("reserved_keywords")) + r")$",
            casefold=str.upper,
        )
    ),
```
[Here in RF04](https://github.com/sqlfluff/sqlfluff/blob/6e15925fc046e1df4841b477df0effc637e3a3ef/src/sqlfluff/rules/references/RF04.py#L85-L96), we fail naked unreserved keywords
```py
if (
    (
        context.segment.is_type("naked_identifier")
        and identifiers_policy_applicable(
            self.unquoted_identifiers_policy,  # type: ignore
            context.parent_stack,
        )
        and (
            context.segment.raw.upper()
            in context.dialect.sets("unreserved_keywords")
        )
    )
```
Naked reserved keywords are not failed by RF04, but they will fail elsewhere due to being unparsable. Should they also fail RF04 anyways for clarity?

So **when you have a quoted unreserved keyword, RF06 will unquote it then RF04 will fail you because it's not quoted**. It should be consistent. Either unreserved keywords should not be unquoted by RF06, or naked unreserved keywords should not fail RF04 (I lean towards the former).

**Comment by ifm-anowaczek** [comment]
I'm willing to make a PR if someone else can weigh in on which direction to take.

**Comment by keraion** [comment]
RF06 allows specifying `sqlfluff:rules:references.quoting.prefer_quoted_keywords` to prevent this mismatch.

```ini
[sqlfluff:rules:references.quoting]
prefer_quoted_keywords = True
```

**Comment by peterbud** [comment]
Closing this as #7089 has been closed as well - there are configuration options to handle such cases.
